cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java

```diff
@@ -0,0 +1,575 @@
+  public void fetchIntermediateCredentials_withCustomUniverseDomain() throws IOException {
+    String universeDomain = "foobar";
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory)
+            .toBuilder()
+            .setUniverseDomain(universeDomain)
+            .build();
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setUniverseDomain(universeDomain)
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    factory.fetchIntermediateCredentials();
+
+    // Verify domain.
+    String url = mockStsTransportFactory.transport.getRequest().getUrl();
+    assertEquals(url, String.format(TOKEN_EXCHANGE_URL_FORMAT, universeDomain));
+  }
+
+  @Test
+  public void fetchIntermediateCredentials_sourceCredentialCannotRefresh_throwsIOException()
+      throws Exception {
+    // Simulate error when refreshing the source credential.
+    mockTokenServerTransportFactory.transport.setError(new IOException());
+
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory);
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    try {
+      factory.fetchIntermediateCredentials(); // Expecting an IOException
+      fail("Should fail as the source credential should not be able to be refreshed.");
+    } catch (IOException e) {
+      assertEquals("Unable to refresh the provided source credential.", e.getMessage());
+    }
+  }
+
+  @Test
+  public void fetchIntermediateCredentials_noExpiresInReturned_copiesSourceExpiration()
+      throws Exception {
+    // Simulate STS not returning expires_in
+    mockStsTransportFactory.transport.setReturnExpiresIn(false);
+
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory);
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    IntermediateCredentials intermediateCredentials = factory.fetchIntermediateCredentials();
+    AccessToken intermediateAccessToken = intermediateCredentials.getIntermediateAccessToken();
+
+    assertEquals(
+        mockStsTransportFactory.transport.getAccessToken(),
+        intermediateAccessToken.getTokenValue());
+
+    // Validate that the expires_in has been copied from the source credential.
+    assertEquals(
+        Objects.requireNonNull(sourceCredentials.getAccessToken()).getExpirationTime(),
```

Comment:
nit: can you bring this `Objects.requireNonNull(sourceCredentials.getAccessToken())` (I think it would be easier to read).

Perhaps with `assertnotnull`?